keps/sig-cluster-lifecycle/wgs/20200108-instance-specific-componentconfig.md

```diff
@@ -0,0 +1,644 @@
+command-line flags to ComponentConfig.
+
+See also issue [#61647](https://github.com/kubernetes/kubernetes/issues/61647).
+
+### Goals
+
+- Provide a clear specification on how to handle instance-specific in
+  ComponentConfig when migrating to a new or adapting an existing
+  ComponentConfig API.
+
+### Non-Goals
+
+- Provide a standard way to continue specifying configuration on the
+  command-line.
+- Redesign APIs/refactor the existing set of parameters. This is about providing
+  a clear migration path for all classes of parameters.
+- Eliminate instance-specific config parameters altogether. This is a special
+  case of the previous item.
+- Make changes to the Dynamic kubelet Config feature or address difficulties in
+  using the kubelet's current ComponentConfig API with Dynamic kubelet Config.
+
+## Proposal
+
+Taking the Kubelet as an example, there are three categories of fields
+(identified by @liggitt):
+1. clearly instance-specific (e.g. `--node-ip`, `--provider-id`,
+   `--hostname-override`)
+2. possibly instance-specific (e.g. `--bind-address`, `--root-dir`)
+3. highly unlikely to be instance-specific (e.g. cloud provider, node lease
+   duration)
+
+This KEP proposes that components have a way to define the same field in
+both instance-specific and shareable configuration files, defining a generic
+merge approach such that the instance-specific field overrides and takes
+precedence over the shareable field.
+
+### Reuse component config Kind for instance-specific config
+
+This KEP proposes a new flag, `--instance-config`, to specify the instance-specific
+configuration file, the instance-specific configuration has the same Kind as the
+sharable one.
+
+ComponentConfig files containing only sharable parameters can continue to be
+shared via a single ConfigMap. Files containing only instance-specific
+parameters can be provided to components via other means, such as the node
+startup scripts or an init container that inserts values from the Pod's Downward
+API.
+
+### Kube-proxy Example
+
+Today, a kube-proxy DaemonSet may be configured as follows to mix sharable and
+instance-specific configuration parameters (using flags only). Note some fields
+and arguments that would be set in production have been removed or modified
+below to improve the clarity of the example.
+
+```yaml
+apiVersion: apps/v1
+kind: DaemonSet
+  name: kube-proxy
+  namespace: kube-system
+spec:
+  template:
+    spec:
+      containers:
+      - name: kube-proxy
+        image: kube-proxy-amd64
+        command:
+        - kube-proxy
+        - --proxy-mode=ipvs
+        - --hostname-override=${NODE_NAME}
+        env:
+          - NODE_NAME
+            valueFrom:
+              fieldRef:
+                fieldPath: spec.nodeName
+```
+
+Below demonstrates how the above may be modified when moving to a shared
+ComponentConfig ConfigMap. Note that the instance-specific parameter is still
+specified via a flag, because the ConfigMap is not appropriate for
```

Comment:
so flags will be mutually exclusive with `--instance-config` and under the hood populate a Config object (patch) in the same way as reading from disc would do?

Reply:
Or, does this proposal eventually mean, let's get rid of the (component-config-bound) flags?

Reply:
Getting rid of component-config-bound flags has been the plan from the beginning, but it's blocked on getting the config to GA.

Reply:
This specific example is showing how things have to be done _before_ instance-specific ComponentConfig is possible.